BootloaderCommonPkg/Library/ElfLib/ElfLib.c

```diff
@@ -0,0 +1,228 @@
+
+#define IS_BOOTABLE_I386_ELF(h) \
+ ( IS_ELF (h)                             && \
+  (h->E_Ident[EI_CLASS]   == ELFCLASS32)  && \
+  (h->E_Ident[EI_DATA]    == ELFDATA2LSB) && \
+  (h->E_Ident[EI_VERSION] == EV_CURRENT)  && \
+ ((h->E_Type              == ET_EXEC)     || (h->E_Type == ET_DYN)) && \
+  (h->E_Machine           == EM_386)      && \
+  (h->E_Version           == EV_CURRENT))
+
+/**
+  Check if the image is a bootable ELF image.
+
+  @param[in]  ImageAddress      Memory address of an image
+
+  @retval     TRUE              Image is a bootable ELF image
+  @retval     FALSE             Not a bootable ELF image
+**/
+BOOLEAN
+EFIAPI
+IsElfImage (
+  IN  VOID  *ImageBase
+  )
+{
+  ELF32_EHDR *Ehdr;
+
+  Ehdr = (ELF32_EHDR *) ImageBase;
+
+  return (BOOLEAN) IS_BOOTABLE_I386_ELF (Ehdr);
+}
+
+/**
+  load ELF image symbol table
+
+  @param[in]      ImageBase   Memory address of ELF image
+  @param[in]      Ehdr        ELF file header
+  @param[in,out]  Eaddr       Start address for ELF symbols
+
+  @retval         End address of loaded sysbols table
+**/
+UINT32
+LoadElfSymtab (
+  IN  VOID   *ImageBase,
+  IN  UINT32  Eaddr
+  )
+{
+  ELF32_EHDR                 *Ehdr;
+  ELF32_SHDR                 *Shdr;
+  UINT32                      Addr;
+  UINT32                      Size;
+  UINT32                      Index;
+
+  // Copy ELF section headers.
+  Ehdr = (ELF32_EHDR *) (UINTN) ImageBase;
+  if (Ehdr->E_Shnum == 0 || Ehdr->E_Shentsize) {
+    return Eaddr;
+  }
+
+  Shdr = (ELF32_SHDR *) (UINTN) Eaddr;
+  Size = Ehdr->E_Shnum * Ehdr->E_Shentsize;
+  CopyMem (Shdr, (VOID *) ((UINT32)ImageBase + Ehdr->E_Shoff), Size);
+  Eaddr += Size;
+
+  // Copy sections that are part of the symbol table.
+  for (Index = 0 ; Index < Ehdr->E_Shnum ; Index += 1) {
+    if (Shdr[Index].Sh_Type != SHT_SYMTAB) {
+      continue;
+    }
+    Addr = ALIGN_UP (Eaddr, Shdr[Index].Sh_Addralign);
+    Size = ALIGN_UP (Shdr[Index].Sh_Size, 4);
+
+    CopyMem ((VOID *)Addr, (VOID *) ((UINT32)ImageBase + Shdr[Index].Sh_Offset), Size);
+    Shdr[Index].Sh_Addr = Addr;
+    Eaddr = Addr + Size;
+  }
+
+  return Eaddr;
+}
+
+/**
+  Load the ELF image to specified address in ELF header.
+
+  This function load ELF image section by section into memory address specified
+  in ELF program header, and also load ELF symbols.
+
+  @param[in]  ImageBase      Memory address of an image
+
+  @retval Image entry point  The entry point of ELF image if load image success
+  @retval NULL               Error with loading ELF image
+**/
+UINT32 *
+LoadElfImage (
+  IN  VOID  *ImageBase
+  )
+{
+  ELF32_EHDR                 *Ehdr;
+  ELF32_PHDR                 *Phdr;
+  UINT32                      Eaddr;
+  UINT32                      Index;
+  UINT32                      Msize;
+  UINT32                      Fsize;
+  UINT32                      Mpos;
+  UINT32                      Ipos;
+
+  //
+  // TODO: Check if image, load region overlap; if yes, move image away.
```

Comment:
Perhaps this can be deferred to another PR, but we should really do some bounds checking to make sure we can actually load the ELF segments into memory at the specified locations. Do we have enough memory? Have the ranges been reserved for something else? Etc.

Reply:
Yes, agree and let's defer this. PE format image relocation as well.